File: framework/languages/dotnet/e2e/RegistrationCodec/Server/Main/Handlers/CodecHandlers.cs
using Google.Protobuf.WellKnownTypes;
using RegistrationCodec.Server.Infrastructure;
using RegistrationCodec.Shared;
using Zlink.Framework.Contracts.Handlers;

namespace RegistrationCodec.Server.Handlers;

internal sealed class JsonEchoRequestHandler(EvidenceStore evidence)
    : IZLinkRequestHandler<JsonEchoReq, EchoReply>
{
    public ValueTask<EchoReply> HandleAsync(JsonEchoReq request, ZLinkRequestContext context, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add($"codec-request|codec=json|value={request.Value}|content={context.ContentType}");
        return ValueTask.FromResult(new EchoReply($"echo:{request.Value}", context.ContentType ?? "<null>"));
    }
}

internal sealed class JsonEchoCommandHandler(EvidenceStore evidence)
    : IZLinkSendHandler<JsonEchoCommand>
{
    public ValueTask HandleAsync(JsonEchoCommand message, ZLinkSendContext context, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add($"codec-command|codec=json|id={message.CommandId}|value={message.Value}|content={context.ContentType}");
        return ValueTask.CompletedTask;
    }
}

internal sealed class ProtobufEchoRequestHandler(EvidenceStore evidence)
    : IZLinkRequestHandler<StringValue, StringValue>
{
    public ValueTask<StringValue> HandleAsync(StringValue request, ZLinkRequestContext context, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add($"codec-request|codec=protobuf|value={request.Value}|content={context.ContentType}");
        return ValueTask.FromResult(new StringValue { Value = $"echo:{request.Value}|content:{context.ContentType}" });
    }
}

internal sealed class ProtobufEchoCommandHandler(EvidenceStore evidence)
    : IZLinkSendHandler<StringValue>
{
    public ValueTask HandleAsync(StringValue message, ZLinkSendContext context, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add($"codec-command|codec=protobuf|value={message.Value}|content={context.ContentType}");
        return ValueTask.CompletedTask;
    }
}

internal sealed class MessagePackEchoRequestHandler(EvidenceStore evidence)
    : IZLinkRequestHandler<PackedEchoReq, PackedEchoReq>
{
    public ValueTask<PackedEchoReq> HandleAsync(PackedEchoReq request, ZLinkRequestContext context, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add($"codec-request|codec=msgpack|value={request.Value}|content={context.ContentType}");
        return ValueTask.FromResult(new PackedEchoReq { Value = $"echo:{request.Value}|content:{context.ContentType}" });
    }
}

internal sealed class MessagePackEchoCommandHandler(EvidenceStore evidence)
    : IZLinkSendHandler<PackedEchoCommand>
{
    public ValueTask HandleAsync(PackedEchoCommand message, ZLinkSendContext context, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        evidence.Add($"codec-command|codec=msgpack|id={message.CommandId}|value={message.Value}|content={context.ContentType}");
        return ValueTask.CompletedTask;
    }
}

File: framework/languages/node/e2e/RegistrationCodec/Server/Main/Handlers/codec-handlers.ts
import { Injectable } from '@nestjs/common';
import type {
  ZLinkRequestContext,
  ZLinkRequestHandler,
  ZLinkSendContext,
  ZLinkSendHandler
} from '@zlink-systems/framework';
import {
  MessagePackEchoReq,
  ProtobufEchoReq,
  type EchoCommand,
  type EchoReply,
  type EchoReq,
  type MessagePackEchoCommand,
  type ProtobufEchoCommand
} from '../../../Shared/messages';
import { EvidenceStore } from '../Infrastructure/evidence-store';

@Injectable()
export class JsonEchoRequestHandler implements ZLinkRequestHandler<EchoReq, EchoReply> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(request: EchoReq, context: ZLinkRequestContext): Promise<EchoReply> {
    this.evidence.add(`codec-request|codec=json|value=${request.value}|content=${context.contentType}`);
    return { value: `echo:${request.value}`, contentType: context.contentType ?? '<null>' };
  }
}

@Injectable()
export class JsonEchoCommandHandler implements ZLinkSendHandler<EchoCommand> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(message: EchoCommand, context: ZLinkSendContext): Promise<void> {
    this.evidence.add(`codec-command|codec=json|id=${message.commandId}|value=${message.value}|content=${context.contentType}`);
  }
}

@Injectable()
export class ProtobufEchoRequestHandler implements ZLinkRequestHandler<EchoReq, ProtobufEchoReq> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(request: EchoReq, context: ZLinkRequestContext): Promise<ProtobufEchoReq> {
    this.evidence.add(`codec-request|codec=protobuf|value=${request.value}|content=${context.contentType}`);
    return new ProtobufEchoReq(`echo:${request.value}|content:${context.contentType ?? '<null>'}`);
  }
}

@Injectable()
export class ProtobufEchoCommandHandler implements ZLinkSendHandler<ProtobufEchoCommand> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(message: ProtobufEchoCommand, context: ZLinkSendContext): Promise<void> {
    this.evidence.add(`codec-command|codec=protobuf|value=${message.value}|content=${context.contentType}`);
  }
}

@Injectable()
export class MessagePackEchoRequestHandler implements ZLinkRequestHandler<EchoReq, MessagePackEchoReq> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(request: EchoReq, context: ZLinkRequestContext): Promise<MessagePackEchoReq> {
    this.evidence.add(`codec-request|codec=msgpack|value=${request.value}|content=${context.contentType}`);
    return new MessagePackEchoReq(`echo:${request.value}|content:${context.contentType ?? '<null>'}`);
  }
}

@Injectable()
export class MessagePackEchoCommandHandler implements ZLinkSendHandler<MessagePackEchoCommand> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(message: MessagePackEchoCommand, context: ZLinkSendContext): Promise<void> {
    this.evidence.add(`codec-command|codec=msgpack|id=${message.commandId}|value=${message.value}|content=${context.contentType}`);
  }
}
